### PR TITLE
Feature/improve duration extraction

### PR DESCRIPTION
**Pull request by cschaefer26**
- Better handling of silent parts and some refactoring.

### CODE REVIEW

File: utils/duration_extractor.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                weight_bottom_right = mat[i + 1, j + 1]
+                row_ind.append(node)
+                col_ind.append(bottom_right_node)
+                data.append(weight_bottom_right)
+
+    adj_mat = coo_matrix((data, (row_ind, col_ind)), shape=(rows * cols, rows * cols))
+    return adj_mat.tocsr()
+
+
+class DurationExtractor:
```
Comment (alexteua) [inline utils/duration_extractor.py:57]: Nice to have this functionality in a class! :)

File: utils/duration_extractor.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+def to_node_index(i: int, j: int, cols: int) -> int:
+    return cols * i + j
+
+
+def from_node_index(node_index: int, cols: int) -> Tuple[int, int]:
+    return node_index // cols, node_index % cols
+
+
+def to_adj_matrix(mat: np.array) -> csr_matrix:
```
Comment (alexteua) [inline utils/duration_extractor.py:19]: Do you think could be a good idea to move them as a class private methods? Or they are planned to be used from outside of the class as well?

Comment (alexteua) [inline utils/duration_extractor.py:19]: Do you think could be a good idea to move them as a class private methods? Or they are planned to be used from outside of the class as well?

File: train_tacotron.py [hunk 3]
```diff
@@ -100,39 +104,50 @@ def create_align_features(model: Tacotron,
...
-            _, _, att_batch = model(batch['x'], batch['mel'])
-        align_score, sharp_score = attention_score(att_batch, batch['mel_len'], r=1)
-        att_batch = np_now(att_batch)
-        seq, att, mel_len, item_id = batch['x'][0], att_batch[0], batch['mel_len'][0], batch['item_id'][0]
-        align_score, sharp_score = float(align_score[0]), float(sharp_score[0])
-        att_score_dict[item_id] = (align_score, sharp_score)
-        durs = dur_extraction_func(seq, att, mel_len)
+            _, _, att_batch = model(x, mel)
+
+        x = batch['x'][0].cpu()
```
Comment (alexteua) [inline train_tacotron.py:128]: Definitely better readability after refactoring :)

Comment (alexteua) [inline train_tacotron.py:128]: Definitely better readability after refactoring :)

File: utils/duration_extractor.py (hunk 2 above)
Comment (cschaefer26) [reply utils/duration_extractor.py]: Agreed, refactored into private methods of the class.

**Review by alexteua — APPROVED** [review]

**Review by tanujjain — APPROVED** [review]

**Pull request merged**